tools/clang/test/CodeGenSPIRV/spv.use-legacy-buffer-matrix-order.hlsl

```diff
@@ -10,26 +10,32 @@ RWByteAddressBuffer buf2;
 [numthreads(64, 1, 1)]
 void main(uint3 tid : SV_DispatchThreadId)
 {
-// CHECK:    [[ptr:%\d+]] = OpAccessChain %_ptr_Uniform_uint %buf %uint_0 [[index_0:%\d+]]
```

Comment:
I do want us to look a little closer at what the optimized code would be here. I don't think we will want to sacrifice performance on 32 and 64 bit loads/stores to fix 16 bit, which might mean having an extra branch in our logic when we're word-aligned. My guess is that if you have enough stuff to load/store, then this:

```
i = byteaddress >> 2
load(i)
i++
load(i)
i++
load(i)
i++
load(i)
...
```

is a lot faster than this:

```
load(byteaddress >> 2)
byteaddress += 4
load(byteaddress >> 2)
byteaddress += 4
load(byteaddress >> 2)
byteaddress += 4
load(byteaddress >> 2)
byteaddress += 4
...
```

and I don't think the optimizer can turn the latter into the former. Even for literal addresses I wouldn't expect it too, but I haven't validated either yet.

Reply:
Okay, so, for literal addresses, for example in this code

```
ByteAddressBuffer buf;

float4 MainPS(float4 pos: SV_Position): SV_Target {
    uint arr[10] = buf.Load<uint[10]>(0);
    return float4(arr[pos.y], 0, 0, 0);
}
```

the code emitted by the optimizer is identical for both before and after this PR. However, you are correct that if you instead `buf.Load<uint[10]>(pos.x)`, the optimizer is unable to remove the shift right instruction.

I've thought of an idea to fix this for word-aligned types, basically just keeping track of whether the latest load/store was from/to a word-aligned type or not, and then only converting from byte address to index and back if it's required. I'll test it out and see how well it works.

Reply:
Okay, check out `BufferAddress` to see my idea for how to solve this. The optimized code for arrays, vectors, and matrices of 32 and 64-bit types is now identical to what it was before (as you can see in this test, for example).